spec/text-editor-presenter-spec.coffee

```diff
@@ -2227,6 +2099,407 @@ describe "TextEditorPresenter", ->
+          expectStateUpdate presenter, -> gutter.hide()
+          decorationState = decorationStateForGutterName(presenter, 'test-gutter')
+          expect(decorationState[decoration1.id]).toBeUndefined()
+          expect(decorationState[decoration2.id]).toBeUndefined()
+          expect(decorationState[decoration3.id]).toBeUndefined()
+
+          # The decorations should return to the original state.
+          expectStateUpdate presenter, -> gutter.show()
+          decorationState = decorationStateForGutterName(presenter, 'test-gutter')
+          expect(decorationState[decoration1.id].top).toBeDefined()
+          expect(decorationState[decoration2.id].top).toBeDefined()
+          expect(decorationState[decoration3.id]).toBeUndefined()
+
+        it "updates when a gutter is added to the editor", ->
+          decorationParams =
+            type: 'gutter'
+            gutterName: 'test-gutter-2'
+            class: 'test-class'
+          marker4 = editor.markBufferRange([[0,0],[1,0]])
+          decoration4 = editor.decorateMarker(marker4, decorationParams)
+          expectStateUpdate presenter, -> editor.addGutter({name: 'test-gutter-2'})
+
+          decorationState = decorationStateForGutterName(presenter, 'test-gutter-2')
+          expect(decorationState[decoration1.id]).toBeUndefined()
+          expect(decorationState[decoration2.id]).toBeUndefined()
+          expect(decorationState[decoration3.id]).toBeUndefined()
+          expect(decorationState[decoration4.id].top).toBeDefined()
+
+        it "does not update when a gutter is destroyed (because it's unnecessary)", ->
+          gutter.destroy()
+          decorationState = decorationStateForGutterName(presenter, 'test-gutter')
+          expect(decorationState[decoration1.id].top).toBeDefined()
+          expect(decorationState[decoration2.id].top).toBeDefined()
+          expect(decorationState[decoration3.id]).toBeUndefined()
+
+          # However, the decoration state of the destroyed gutter should be
+          # cleared whenever ::updateCustomGutterDecorationState is next called.
+          expectStateUpdate presenter, -> editor.addGutter({name: 'a-different-gutter'})
+          decorationState = decorationStateForGutterName(presenter, 'test-gutter')
+          expect(decorationState).toBeUndefined()
+
+        it "updates when editor lines are folded", ->
+          oldDimensionsForDecoration1 =
+            top: lineHeight * marker1.getScreenRange().start.row
+            height: lineHeight * marker1.getScreenRange().getRowCount()
+          oldDimensionsForDecoration2 =
+            top: lineHeight * marker2.getScreenRange().start.row
+            height: lineHeight * marker2.getScreenRange().getRowCount()
+
+          # Based on the contents of sample.js, this should affect all but the top
+          # part of decoration1.
+          expectStateUpdate presenter, -> editor.foldBufferRow(0)
```

Comment:
Might be nice to check a smaller fold to confirm that decorations after the fold are shifted upward appropriately, proving that you're following screen positions.